transports/quic/src/muxer.rs

```diff
@@ -89,6 +143,93 @@ impl QuicMuxer {
+    type Error = Error;
+
+    fn poll_address_change(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<libp2p_core::Multiaddr, Self::Error>> {
+        self.inner.lock().poll_connection(cx);
+        // TODO
+        Poll::Pending
+    }
+
+    fn poll_inbound(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Self::Substream, Self::Error>> {
+        let mut inner = self.inner.lock();
+        inner.poll_connection(cx);
+        if let Some(substream_id) = inner.connection.pop_incoming_substream() {
+            inner.substreams.insert(substream_id, Default::default());
+            let substream = Substream::new(substream_id, self.inner.clone());
+            Poll::Ready(Ok(substream))
+        } else {
+            inner.poll_event_waker = Some(cx.waker().clone());
+            Poll::Pending
+        }
+    }
+
+    fn poll_outbound(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Self::Substream, Self::Error>> {
+        let mut inner = self.inner.lock();
+        inner.poll_connection(cx);
+        if let Some(substream_id) = inner.connection.pop_outgoing_substream() {
+            inner.substreams.insert(substream_id, Default::default());
+            let substream = Substream::new(substream_id, self.inner.clone());
+            Poll::Ready(Ok(substream))
+        } else {
+            inner.pending_substreams.push_back(cx.waker().clone());
+            Poll::Pending
+        }
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let mut inner = self.inner.lock();
+        inner.poll_connection(cx);
+
+        if inner.connection.connection.is_drained() {
+            return Poll::Ready(Ok(()));
+        }
+
+        if inner.substreams.is_empty() {
+            let connection = &mut inner.connection;
+            if !connection.connection.is_closed() {
+                connection.close();
+                if let Some(waker) = inner.poll_event_waker.take() {
+                    waker.wake();
+                }
+            } else {
+            }
+            while let Poll::Ready(event) = inner.connection.poll_event(cx) {
+                if let ConnectionEvent::ConnectionLost(_) = event {
+                    return Poll::Ready(Ok(()));
+                }
+            }
+        } else {
+            for substream in inner.substreams.clone().keys() {
+                if let Err(e) = inner.connection.shutdown_substream(*substream) {
+                    tracing::error!("substream finish error on muxer close: {}", e);
+                }
+            }
+        }
+
+        // Register `cx.waker()` as being woken up if the connection closes.
+        inner.poll_close_waker = Some(cx.waker().clone());
+
+        Poll::Pending
+    }
+}
+
+impl fmt::Debug for QuicMuxer {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("QuicMuxer").finish()
```

Comment:
We could print a few things here, like how many substreams we have open for example. Not sure if it is necessary.

Reply:
We can just derive it for `QuicMuxerInner` and thus also for `QuixMuxer`.